language/ast.py

```diff
@@ -0,0 +1,779 @@
+from langkit.dsl import ASTNode, Field, abstract  # type: ignore
+
+
+@abstract
+class RFLXNode(ASTNode):
+    """
+    Root node class for the RecordFlux DSL language
+    """
+
+
+class NullID(RFLXNode):
+    """
+    "null" identifier
+    """
+
+
+class UnqualifiedID(NullID):
+    """
+    Simple, unqualified identifiers, i.e. identifiers without a package part (e.g. "Foo")
+    """
+
+    token_node = True
+
+
+class ID(NullID):
+    """
+    Qualified identifiers which may optionally have a package part (e.g. "Pkg::Foo", "Foo")
+    """
+
+    package = Field(type=UnqualifiedID)
+    name = Field(type=UnqualifiedID)
+
+
+@abstract
+class Type(RFLXNode):
+    """
+    Base class for type declarations (enums, integers, messages, arrays, derivations)
+    """
+
+
+@abstract
+class TypeDef(RFLXNode):
+    """
+    Base class for type definitions
+    """
+
+
+class TypeSpec(Type):
+    """
+    Type specification (type Foo is ...)
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    definition = Field(type=TypeDef)
+
+
+@abstract
+class Expr(RFLXNode):
+    """
+    Base class for expressions
+    """
+
+
+class Op(RFLXNode):
+    """
+    Operators for expressions
+    """
+
+    enum_node = True
+    alternatives = [
+        "pow",
+        "mul",
+        "div",
+        "add",
+        "sub",
+        "mod",
+        "eq",
+        "neq",
+        "le",
+        "lt",
+        "gt",
+        "ge",
+        "and",
+        "or",
+        "in",
+        "notin",
+    ]
+
+
+class Negation(Expr):
+    """
+    Negation
+    """
+
+    data = Field(type=Expr)
+
+
+class BinOp(Expr):
+    """
+    Binary operation
+    """
+
+    left = Field(type=Expr)
+    op = Field(type=Op)
+    right = Field(type=Expr)
+
+
+class ParenExpression(Expr):
+    """
+    Parenthesized expression
+    """
+
+    data = Field(type=Expr)
+
+
+class RefinementSpec(Type):
+    """
+    Refinement specification (for Message use (Field => Inner_Type))
+    """
+
+    pdu = Field(type=ID)
+    field = Field(type=UnqualifiedID)
+    sdu = Field(type=ID)
+    condition = Field(type=Expr)
+
+
+@abstract
+class TypeDecl(RFLXNode):
+    """
+    Base class for type declarations
+    """
```

Comment:
Why are there two base classes for type declarations? `class Type(RFLXNode)` has the same documentation.

Reply:
Changed in 669e85c.